program-metadata/program/src/processor.rs

```diff
@@ -0,0 +1,449 @@
+        borsh::try_from_slice_unchecked,
+        entrypoint::ProgramResult,
+        msg,
+        pubkey::Pubkey,
+    },
+};
+
+pub fn process_instruction(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    input: &[u8],
+) -> ProgramResult {
+    let instruction = MetadataInstruction::try_from_slice(input)?;
+
+    match instruction {
+        MetadataInstruction::CreateMetadataEntry {
+            name,
+            value,
+            hashed_name,
+        } => {
+            msg!("Instruction: Create Metadata Entry");
+            process_create_metadata_entry(program_id, accounts, name, value, hashed_name)
+        }
+        MetadataInstruction::UpdateMetadataEntry { value } => {
+            msg!("Instruction: Update Metadata Entry");
+            process_update_metadata_entry(program_id, accounts, value)
+        }
+        MetadataInstruction::DeleteMetadataEntry => {
+            msg!("Instruction: Delete Metadata Entry");
+            process_delete_metadata_entry(program_id, accounts)
+        }
+        MetadataInstruction::CreateVersionedIdl {
+            effective_slot,
+            idl_url,
+            idl_hash,
+            source_url,
+            hashed_name,
+        } => {
+            msg!("Instruction: Create Versioned Idl");
+            process_create_versioned_idl(
+                program_id,
+                accounts,
+                effective_slot,
+                idl_url,
+                idl_hash,
+                source_url,
+                hashed_name,
+            )
+        }
+        MetadataInstruction::UpdateVersionedIdl {
+            idl_url,
+            idl_hash,
+            source_url,
+        } => {
+            msg!("Instruction: Update Versioned IDL");
+            process_update_versioned_idl(program_id, accounts, idl_url, idl_hash, source_url)
+        }
+    }
+}
+
+pub fn process_create_metadata_entry(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    name: String,
+    value: String,
+    hashed_name: [u8; 32],
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+    let class_account_info = next_account_info(account_info_iter)?;
+    let name_account_info = next_account_info(account_info_iter)?;
+    let target_program_info = next_account_info(account_info_iter)?;
+    let target_program_program_data_info = next_account_info(account_info_iter)?;
+    let target_program_authority_info = next_account_info(account_info_iter)?;
+    let payer_account_info = next_account_info(account_info_iter)?;
+    let system_account_info = next_account_info(account_info_iter)?;
+    let rent_sysvar_info = next_account_info(account_info_iter)?;
+    let name_service_info = next_account_info(account_info_iter)?;
+
+    if name.len() > MAX_NAME_LENGTH {
+        return Err(MetadataError::NameTooLong.into());
+    }
+
+    if value.len() > MAX_VALUE_LENGTH {
+        return Err(MetadataError::ValueTooLong.into());
+    }
```

Comment:
It would be more flexible to add limit the sum of these two lengths. That way devs could use longer values with shorter names